### PR TITLE
Use variant-schema-compiler for some GVariant code

### PR DESCRIPTION
**Pull request by alexlarsson**
This replaces some use of g_variant_get_* with generated code based on a IDL description of the types.

The generated code is a lot more efficient than the individual g_variant calls, as well as easier to read due to the use of types and field names instead of magic offsets. On the other hand, this is not necessarily code that needs to be very efficient, and there is a risk in any change like this.

Is this something we want to use, and over time extend to more code?

### CODE REVIEW

**Comment by matthiasclasen** [comment]
I think it would be a shame not to use it, now that you've done the work.

**Comment by mwleeds** [comment]
What does IDL stand for in this context?

**Comment by alexlarsson** [comment]
IDL == interface description language

**Comment by alexlarsson** [comment]
There is still a bunch of summary and cache lookup stuff I want to convert.

**Comment by mwleeds** [comment]
I have more urgent things to work on right now but I can review this at some point.

**Comment by mwleeds** [comment]
> There is still a bunch of summary and cache lookup stuff I want to convert.

In this PR or as follow-up?

File: common/flatpak-dir.c [hunk 1]
```diff
@@ -2278,169 +2276,174 @@ flatpak_dir_repo_lock (FlatpakDir   *self,
...
-    g_hash_table_insert (content_rating,
-                         (gpointer) g_intern_string (id),
-                         (gpointer) g_intern_string (val));
+          g_hash_table_insert (content_rating,
+                               (gpointer) g_intern_string (var_ratings_entry_get_key (entry)),
+                               (gpointer) g_intern_string (var_ratings_entry_get_value (entry)));
+        }
+    }
 
   return g_steal_pointer (&content_rating);
```
Comment (mwleeds) [inline common/flatpak-dir.c]: Looks like we're returning an empty hash table in case the info is missing, whereas previously we returned NULL in that case. Not sure if that's going to cause problems.

File: common/flatpak-dir.c [hunk 2]
```diff
@@ -2451,21 +2454,17 @@ flatpak_deploy_data_get_appdata_content_rating (GVariant *deploy_data)
  * Returns: (array length=length zero-terminated=1) (transfer container): an array of constant strings
```
Comment (mwleeds) [inline common/flatpak-dir.c]: `length=length` looks like a copy-paste error, though not the fault of this commit

File: common/flatpak-dir.c [hunk 3]
```diff
@@ -11422,11 +11425,7 @@ flatpak_dir_list_all_remote_refs (FlatpakDir         *self,
 
   exts = var_summary_get_metadata (summary);
 
-  collection_id = NULL;
-  if (var_metadata_lookup (exts, "ostree.summary.collection-id", NULL, &v) &&
-      var_variant_is_type (v, G_VARIANT_TYPE_STRING))
-    collection_id = var_variant_get_string (v);
-
+  collection_id = var_metadata_lookup_string (exts, "ostree.summary.collection-id", NULL);
```
Comment (mwleeds) [inline common/flatpak-dir.c]: This change doesn't really belong in this commit right?

File: common/Makefile.am.inc [hunk 4]
```diff
@@ -150,6 +156,7 @@ libflatpak_common_la_SOURCES = \
 	common/flatpak-utils-private.h \
 	common/flatpak-utils.c \
 	common/valgrind-private.h \
+	common/flatpak-variant-private.h \
```
[the remark this answers is not in the document]
Comment (alexlarsson) [reply common/Makefile.am.inc]: Yes, i'd like to ship this in the tarball so people building flatpak don't have to have pyparsing installed.

File: common/flatpak-dir.c [hunk 5]
```diff
@@ -11422,29 +11414,21 @@ flatpak_dir_list_all_remote_refs (FlatpakDir         *self,
...
-
-      cache = g_variant_get_child_value (xa_cache, 0);
-      n = g_variant_n_children (cache);
+      n = var_cache_get_length (cache);
       for (i = 0; i < n; i++)
         {
-          g_autoptr(GVariant) child = NULL;
-          g_autoptr(GVariant) cur_v = NULL;
           g_autoptr(FlatpakCollectionRef) coll_ref = NULL;
+          VarCacheEntryRef entry;
```
[the remark this answers is not in the document]
Comment (alexlarsson) [reply common/flatpak-dir.c]: An Entry is the dict entry, i.e. the thing with both key and value in it. The CacheData is the value only.

File: common/flatpak-dir.c [hunk 6]
```diff
@@ -11358,41 +11358,33 @@ flatpak_dir_remote_has_ref (FlatpakDir *self,
...
-  while ((value = g_variant_iter_next_value (&ref_iter)) != NULL)
+  len = var_ref_map_get_length (ref_map);
+  for (i = 0; i < len; i++)
     {
-      /* helper for being able to auto-free the value */
-      g_autoptr(GVariant) child = value;
-      const char *ref_name = NULL;
-
-      g_variant_get_child (child, 0, "&s", &ref_name);
-      if (ref_name == NULL)
```
[the remark this answers is not in the document]
Comment (alexlarsson) [reply common/flatpak-dir.c]: String elements can't be NULL in a variant. g_variant_get_child would only ever return NULL here if child 0 was not a string, but at this point we already asserted the type.

File: common/flatpak-dir.c [hunk 7]
```diff
@@ -11406,11 +11398,11 @@ flatpak_dir_list_all_remote_refs (FlatpakDir         *self,
                                   GError            **error)
 {
   g_autoptr(GHashTable) ret_all_refs = NULL;
-  g_autoptr(GVariant) ref_map = NULL;
-  g_autoptr(GVariant) exts = NULL;
-  g_autoptr(GVariant) collection_map = NULL;
   const gchar *collection_id;
-  GVariantIter iter;
+  VarSummaryRef summary;
+  VarMetadataRef exts;
```
[the remark this answers is not in the document]
Comment (alexlarsson) [reply common/flatpak-dir.c]: It doesn't really have a name, its just a dict that lets you shove random extra stuff in the summary. I guess i meant extensions? I'm not sure that is a great name either though.

File: common/Makefile.am.inc [hunk 8]
```diff
@@ -75,6 +75,12 @@ nodist_libflatpak_common_base_la_SOURCES = \
 BUILT_SOURCES += $(nodist_libflatpak_common_base_la_SOURCES)
 CLEANFILES += $(nodist_libflatpak_common_base_la_SOURCES)
 
+common/flatpak-variant-private.h: variant-schema-compiler/variant-schema-compiler  data/flatpak-variants.gv
+	$(AM_V_GEN) variant-schema-compiler/variant-schema-compiler --outfile common/flatpak-variant-private.h --prefix var data/flatpak-variants.gv
```
[the remark this answers is not in the document]
Comment (alexlarsson) [reply common/Makefile.am.inc]: The generated types are bit weird, not being typical GObjects and stuff, and some may conflict with other more regular types, so I picked a different prefix.

File: common/flatpak-dir.c (hunk 3 above)
Comment (alexlarsson) [reply common/flatpak-dir.c]: Yeah, it just snuck in due to new generated APIs being available

**Comment by alexlarsson** [comment]
Rebased to master and fixes up some of the comments. I still want to do some more work before we land this. Also, i'm considering moving the generator to gnome git.
